src/Communicator/Services/Implementations/SmsService.cs

```diff
@@ -1,18 +1,211 @@
+    {
+        _channel = channel;
+        _smsConfiguration = GetSmsConfigurationByChannel();
+        _httpClient = GenerateProviderHttpClient();
+    }
+
+    private SmsConfiguration GetSmsConfigurationByChannel()
     {
-        throw new NotImplementedException();
+        return options.SmsConfigurations?.FirstOrDefault(x => x.Key == _channel).Value
+               ?? throw new ArgumentException("No valid provider with given channel");
     }
 
-    public Task SendBulkAsync(List<SmsMessage> smsList, CancellationToken cancellationToken = default)
+    private static SmsProviders GetSmsProvider(SmsConfiguration smsConfiguration)
     {
-        throw new NotImplementedException();
+        Enum.TryParse(typeof(SmsProviders), smsConfiguration.Provider, out var providerValue);
+
+        if (providerValue is null)
+        {
+            throw new Exception("Wrong provider");
+        }
+
+        var provider = (SmsProviders)providerValue;
+
+        return provider;
     }
 
+    private HttpClient GenerateProviderHttpClient()
+    {
+        var client = httpClientFactory.CreateClient(_channel);
+
+        var provider = GetSmsProvider(_smsConfiguration);
+
+        var properties = _smsConfiguration.Properties;
+
+        switch (provider)
+        {
+            case SmsProviders.Dexatel:
+                client.DefaultRequestHeaders.Add("X-Dexatel-Key", properties["X-Dexatel-Key"]);
+                break;
+
+            case SmsProviders.Twilio:
+                client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic",
+                    Convert.ToBase64String(
+                        Encoding.ASCII.GetBytes($"{properties["SID"]}:{properties["AUTH_TOKEN"]}")));
+                break;
+        }
+
+        return client;
+    }
+
+    private async Task<List<GeneralSmsResponse>> SendSmsAsync(SmsMessage smsMessage,
+        CancellationToken cancellationToken)
+    {
+        var provider = GetSmsProvider(_smsConfiguration);
+
+        switch (provider)
+        {
+            case SmsProviders.Dexatel:
+                return await SendSmsViaDexatelAsync(smsMessage, cancellationToken);
+
+            case SmsProviders.Twilio:
+                return await SendSmsViaTwilioAsync(smsMessage, cancellationToken);
+
+            default:
+                throw new ArgumentOutOfRangeException();
+        }
+    }
+
+    private async Task<List<GeneralSmsResponse>> SendSmsViaDexatelAsync(SmsMessage smsMessage,
+        CancellationToken cancellationToken)
+    {
+        var request = new DexatelSmsSendRequest
+        {
+            Data = new DexatelSmsSendRequestData
+            {
+                From = _smsConfiguration.From,
+                To = smsMessage.Recipients.MakeDistinct(),
+                Text = smsMessage.Message,
+                Channel = "SMS"
+            }
+        };
+
+        var response = await _httpClient.PostAsync(
```

Comment:
separate method for _httpClient.PostAsync()

Reply:
done